src/app.test.js

```diff
@@ -40,16 +41,36 @@ const addTodo = (todoList, text, completed) => {
   }
 };
 
-test("renders empty todo list when starts", () => {
+const logInUser = (userName) => {
+  const userInput = screen.getByPlaceholderText("Please enter username");
+  const setUserButton = screen.getByText("Set User");
+
+  userEvent.type(userInput, userName);
+  userEvent.click(setUserButton);
+};
+test("wont allow to add new todo until user is not set", () => {
+  // act
+  const todoList = screen.getByTestId("todo-list");
+  addTodo(todoList, "Take wife for a walk!");
+
+  // assert
+  expect(
+    queryByDisplayValue(todoList, "Take wife for a walk!")
+  ).not.toBeInTheDocument();
+});
+
+test("renders empty todo list when new user is logs in", () => {
   // setup
+  logInUser("Attila");
```

Comment:
this has become a duplicate now:

```js
logInUser("Attila");
```

as it's been repeated in almost every test case

let's use the parameter default value in `logInUser` instead and only pass in the username when the scenario requires a different one

Reply:
btw the "login" step is actually an "act" rather then part of the "setup" phase

Reply:
once it's done, I'd have a think about if we could move this to `beforeEach` as almost all test cases need it - apart from the first one which is checking that without logging in the user cannot add a todo

the question is how can we let the application "log out" the current user? :)